flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/BinaryString.java

```diff
@@ -654,531 +498,211 @@ private BinaryString trimSlow() {
-	 * @return the position of the first occurence of substring. Return -1 if not found.
-	 */
-	public int indexOf(BinaryString subStr, int start) {
-		ensureMaterialized();
-		subStr.ensureMaterialized();
-		if (subStr.sizeInBytes == 0) {
-			return 0;
-		}
-		if (inFirstSegment()) {
-			// position in byte
-			int byteIdx = 0;
-			// position is char
-			int charIdx = 0;
-			while (byteIdx < sizeInBytes && charIdx < start) {
-				byteIdx += numBytesForFirstByte(getByteOneSegment(byteIdx));
-				charIdx++;
-			}
-			do {
-				if (byteIdx + subStr.sizeInBytes > sizeInBytes) {
-					return -1;
-				}
-				if (SegmentsUtil.equals(segments, offset + byteIdx,
-					subStr.segments, subStr.offset, subStr.sizeInBytes)) {
-					return charIdx;
-				}
-				byteIdx += numBytesForFirstByte(getByteOneSegment(byteIdx));
-				charIdx++;
-			} while (byteIdx < sizeInBytes);
-
-			return -1;
-		} else {
-			return indexOfSlow(subStr, start);
-		}
+	private boolean matchAtOneSeg(final BinaryString s, int pos) {
+		return s.sizeInBytes + pos <= sizeInBytes && pos >= 0 &&
+			segments[0].equalTo(s.segments[0], offset + pos, s.offset, s.sizeInBytes);
 	}
 
-	private int indexOfSlow(BinaryString subStr, int start) {
-		// position in byte
-		int byteIdx = 0;
-		// position is char
-		int charIdx = 0;
-		int segSize = segments[0].size();
-		SegmentAndOffset index = firstSegmentAndOffset(segSize);
-		while (byteIdx < sizeInBytes && charIdx < start) {
-			int charBytes = numBytesForFirstByte(index.value());
-			byteIdx += charBytes;
-			charIdx++;
-			index.skipBytes(charBytes, segSize);
-		}
-		do {
-			if (byteIdx + subStr.sizeInBytes > sizeInBytes) {
-				return -1;
-			}
-			if (SegmentsUtil.equals(segments, offset + byteIdx,
-				subStr.segments, subStr.offset, subStr.sizeInBytes)) {
-				return charIdx;
-			}
-			int charBytes = numBytesForFirstByte(index.segment.get(index.offset));
-			byteIdx += charBytes;
-			charIdx++;
-			index.skipBytes(charBytes, segSize);
-		} while (byteIdx < sizeInBytes);
-
-		return -1;
+	private boolean matchAtVarSeg(final BinaryString s, int pos) {
+		return s.sizeInBytes + pos <= sizeInBytes && pos >= 0 &&
+			SegmentsUtil.equals(segments, offset + pos, s.segments, s.offset, s.sizeInBytes);
 	}
 
-	/**
-	 * Reverse each character in current string.
-	 *
-	 * @return a new string which character order is reverse to current string.
-	 */
-	public BinaryString reverse() {
-		ensureMaterialized();
-		if (inFirstSegment()) {
-			byte[] result = new byte[this.sizeInBytes];
-			// position in byte
-			int byteIdx = 0;
-			while (byteIdx < sizeInBytes) {
-				int charBytes = numBytesForFirstByte(getByteOneSegment(byteIdx));
-				segments[0].get(
-					offset + byteIdx,
-					result,
-					result.length - byteIdx - charBytes,
-					charBytes);
-				byteIdx += charBytes;
-			}
-			return BinaryString.fromBytes(result);
-		} else {
-			return reverseSlow();
-		}
+	BinaryString copyBinaryStringInOneSeg(int start, int end) {
```

Comment:
the end index is inclusive, which seems to be different from Java convention

Reply:
I'll change to len